File: frontend/src/types/question.ts
export type QuestionType = "single" | "multiple" | "judge";

export type QuestionOption = {
  id: number;
  label: string;
  content: string;
  is_correct: boolean;
  sort_order: number;
};

export type QuestionOptionPayload = Omit<QuestionOption, "id">;

export type AdminQuestion = {
  id: number;
  question_type: QuestionType | string;
  stem: string;
  analysis?: string | null;
  category_1?: string | null;
  category_2?: string | null;
  difficulty?: string | null;
  score: number;
  status: string;
  source?: string | null;
  source_no?: string | null;
  remark?: string | null;
  options: QuestionOption[];
};

export type PracticeQuestionOption = Omit<QuestionOption, "is_correct">;

export type PracticeQuestion = Omit<AdminQuestion, "analysis" | "options"> & {
  options: PracticeQuestionOption[];
};

export type Question = AdminQuestion;

export type QuestionPayload = Omit<AdminQuestion, "id" | "options"> & {
  options: QuestionOptionPayload[];
};

export type PracticeAnswerResult = {
  practice_answer_id: number;
  question_id: number;
  selected_answer: string;
  score: number;
  is_correct: boolean;
  correct_answer: string;
  analysis?: string | null;
  option_comparison: PracticeOptionComparison[];
};

export type PracticeOptionComparison = {
  label: string;
  content: string;
  selected: boolean;
  correct: boolean;
};

export type PracticeAnswerHistory = {
  practice_answer_id: number;
  selected_answer: string;
  is_correct: boolean;
  practiced_at: string;
};

export type PracticeWrongQuestion = {
  question_id: number;
  question_type: QuestionType | string;
  stem: string;
  category_1?: string | null;
  category_2?: string | null;
  status: string;
  correct_answer: string;
  analysis?: string | null;
  incorrect_count: number;
  total_attempts: number;
  mastered: boolean;
  latest_practiced_at: string;
  history: PracticeAnswerHistory[];
  options: PracticeOptionComparison[];
};

File: backend/app/models/question.py
from decimal import Decimal

from sqlalchemy import (
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.base import TimestampMixin


class Question(TimestampMixin, Base):
    __tablename__ = "question"
    __table_args__ = (
        Index("ix_question_category", "category_1", "category_2"),
        Index("ix_question_type_status", "question_type", "status"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    question_type: Mapped[str] = mapped_column(String(20), nullable=False)
    stem: Mapped[str] = mapped_column(Text, nullable=False)
    analysis: Mapped[str | None] = mapped_column(Text)
    category_1: Mapped[str | None] = mapped_column(String(100))
    category_2: Mapped[str | None] = mapped_column(String(100))
    difficulty: Mapped[str | None] = mapped_column(String(50))
    score: Mapped[Decimal] = mapped_column(
        Numeric(8, 2), nullable=False, default=1, server_default="1"
    )
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="active")
    source: Mapped[str | None] = mapped_column(String(200))
    source_no: Mapped[str | None] = mapped_column(String(100))
    remark: Mapped[str | None] = mapped_column(Text)

    options = relationship(
        "QuestionOption", back_populates="question", cascade="all, delete-orphan"
    )
    practice_answers = relationship("PracticeAnswer", back_populates="question")


class QuestionOption(TimestampMixin, Base):
    __tablename__ = "question_option"
    __table_args__ = (
        UniqueConstraint("question_id", "label", name="uq_question_option_label"),
    )

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    question_id: Mapped[int] = mapped_column(
        ForeignKey("question.id", ondelete="CASCADE"), nullable=False, index=True
    )
    label: Mapped[str] = mapped_column(String(10), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    is_correct: Mapped[bool] = mapped_column(nullable=False, default=False)
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, default=0)

    question = relationship("Question", back_populates="options")
